src/mscorlib/src/System/String.Manipulation.cs

```diff
@@ -1439,92 +1454,81 @@ private int MakeSeparatorList(ReadOnlySpan<char> separators, int[] sepList)
                             if (IsCharBitSet(charMap, (byte)c) && IsCharBitSet(charMap, (byte)(c >> 8)) &&
                                 separators.Contains(c))
                             {
-                                sepList[foundCount++] = i;
+                                sepListBuilder.Append(i);
                             }
                         }
                     }
                     break;
             }
-
-            return foundCount;
         }
 
-        //--------------------------------------------------------------------
-        // This function returns number of the places within baseString where
-        // instances of the separator string occurs.
-        // Args: separator  -- the separator
-        //       sepList    -- an array of ints for split string indicies.
-        //--------------------------------------------------------------------
-        private unsafe int MakeSeparatorList(string separator, int[] sepList)
+        /// <summary>
+        /// Uses ValueListBuilder to create list that holds indexes of separators in string.
+        /// </summary>
+        /// <param name="separator">separator string</param>
+        /// <param name="sepListBuilder"><see cref="ValueListBuilder{T}"/> to store indexes</param>
+        /// <returns></returns>
+        private unsafe void MakeSeparatorList(string separator, ref ValueListBuilder<int> sepListBuilder)
         {
-            Debug.Assert(!string.IsNullOrEmpty(separator), "!string.IsNullOrEmpty(separator)");
+            Debug.Assert(!IsNullOrEmpty(separator), "!string.IsNullOrEmpty(separator)");
 
-            int foundCount = 0;
-            int sepListCount = sepList.Length;
             int currentSepLength = separator.Length;
 
             fixed (char* pwzChars = &_firstChar)
```

Comment:
This unsafe code should not be necessary now that you have simplified the loop condition. `for (int i = 0; i < s.Length; i++) { s[i]` is a pattern that the JIT bounds-check elimination is pretty good at recognizing.